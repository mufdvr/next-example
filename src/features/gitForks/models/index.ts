export * from './fork'
export * from './user'
