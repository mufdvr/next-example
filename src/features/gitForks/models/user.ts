export interface IUser {
  id: number
  login: string
  avatarUrl: string
}

export class User implements IUser {
  public id: number
  public login: string
  public avatarUrl: string

  constructor(props: any = {}) {
    this.id = +props.id || -1
    this.login = props.login || ''
    this.avatarUrl = props.avatarUrl || ''
  }
}
