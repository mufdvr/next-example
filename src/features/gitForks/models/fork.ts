import humps from 'humps'

import { User, IUser } from './user'

export interface IForksRequest {
  userName: string
  repoName: string
  page?: number
  isServer?: boolean
}

export interface IFork {
  id: number
  fullName: string
  htmlUrl: string
  stargazersCount: number
  owner: IUser
}

export class Fork implements IFork {

  public static create(dto: {}): IFork {
    const fork: any = humps.camelizeKeys(dto)
    fork.owner = new User(fork.owner)
    return new Fork(fork)
  }

  public id: number
  public fullName: string
  public htmlUrl: string
  public stargazersCount: number
  public owner: IUser

  constructor(props: any = {}) {
    this.id = +props.id || -1
    this.fullName = props.fullName || ''
    this.htmlUrl = props.htmlUrl || ''
    this.stargazersCount = +props.stargazersCount || 0
    this.owner = props.owner || ''
  }
}
