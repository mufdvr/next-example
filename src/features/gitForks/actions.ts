import { createAsyncAction } from 'typesafe-actions'
import { IForksRequest, IFork } from './models'

import * as types from './actionTypes'

export const fetchRepos = createAsyncAction(types.FETCH_REPOS_REQUEST, types.FETCH_REPOS_SUCCESS, types.FETCH_REPOS_FAILURE)<
  string,
  string[],
  IRequestError
>()

export const fetchForks = createAsyncAction(types.FETCH_FORKS_REQUEST, types.FETCH_FORKS_SUCCESS, types.FETCH_FORKS_FAILURE)<
  IForksRequest,
  IFork[],
  IRequestError
>()
