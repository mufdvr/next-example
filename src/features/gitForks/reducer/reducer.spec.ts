import forksReducer, { IForksState } from './forks'
import reposReducer, { IReposState } from './repos'
import * as actions from '../actions'
import { Fork } from '../models'

const error = {
  message: 'Not found',
  status: 404,
}

describe('ForksSearching Stories', () => {
  const getInitialState = (initial?: Partial<IForksState>) => forksReducer(initial as IForksState, {} as any)
  const initialState = getInitialState()

  it('should match a snapshot', () => {
    expect(initialState).toMatchSnapshot()
  })

  it('forks request fetching true', done => {
    const state = forksReducer(initialState, actions.fetchForks.request({ userName: '', repoName: '' }))
    expect(state).toEqual({
      ...initialState,
      fetching: true,
    })
    done()
  })

  it('forks fetching success', done => {
    const forks = [new Fork()]
    expect(initialState.forks).toHaveLength(0)
    const state = forksReducer(initialState, actions.fetchForks.success(forks))
    expect(state).toEqual({
      ...initialState,
      forks,
    })
    done()
  })

  it('forks fetching error', done => {
    const state = forksReducer(initialState, actions.fetchForks.failure(error))
    expect(state.forks).toHaveLength(0)
    expect(state).toEqual({
      ...initialState,
      error,
    })
    done()
  })
})

describe('Get Repos Stories', () => {
  const getInitialState = (initial?: Partial<IForksState>) => reposReducer(initial as IReposState, {} as any)
  const initialState = getInitialState()

  it('should match a snapshot', () => {
    expect(initialState).toMatchSnapshot()
  })

  it('repos request fetching true', done => {
    const state = reposReducer(initialState, actions.fetchRepos.request('wtf'))
    expect(state).toEqual({
      ...initialState,
      fetching: true,
    })
    done()
  })

  it('repos fetching success', done => {
    const repos = ['repo-name']
    expect(initialState.repos).toHaveLength(0)
    const state = reposReducer(initialState, actions.fetchRepos.success(repos))
    expect(state).toEqual({
      ...initialState,
      repos,
    })
    done()
  })

  it('repos fetching error', done => {
    const state = reposReducer(initialState, actions.fetchRepos.failure(error))
    expect(state.repos).toHaveLength(0)
    expect(state).toEqual({
      ...initialState,
      error,
    })
    done()
  })
})
