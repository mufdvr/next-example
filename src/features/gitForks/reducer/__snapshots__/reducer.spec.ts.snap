// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForksSearching Stories should match a snapshot 1`] = `
Object {
  "error": Object {
    "message": "",
    "status": 0,
  },
  "fetching": false,
  "forks": Array [],
}
`;

exports[`Get Repos Stories should match a snapshot 1`] = `
Object {
  "error": Object {
    "message": "",
    "status": 0,
  },
  "fetching": false,
  "repos": Array [],
}
`;
